examples/step-65/step-65.cc

```diff
@@ -0,0 +1,389 @@
+    constraints.distribute(solution);
+    std::cout << "   Number of solver iterations:  "
+              << solver_control.last_step() << std::endl;
+  }
+
+
+
+  template <int dim>
+  void Step64<dim>::output_results(const Mapping<dim> &mapping) const
+  {
+    {
+      TimerOutput::Scope scope(timer, "Write output");
+      Timer              time;
+      DataOut<dim>       data_out;
+      data_out.attach_dof_handler(dof_handler);
+      data_out.add_data_vector(solution, "solution");
+
+      DataOutBase::VtkFlags flags;
+      flags.write_higher_order_cells = true;
+      data_out.set_flags(flags);
+
+      Vector<double> material_ids(triangulation.n_active_cells());
+      for (const auto &cell : triangulation.active_cell_iterators())
+        material_ids[cell->active_cell_index()] = cell->material_id();
+      data_out.add_data_vector(material_ids, "material_ids");
+
+      data_out.build_patches(mapping,
+                             fe.degree,
+                             DataOut<dim>::curved_inner_cells);
+
+      std::ofstream file(
+        ("solution-" +
+         std::to_string(triangulation.n_global_levels() - 10 + 2 * dim) +
+         ".vtu")
+          .c_str());
+      data_out.write_vtu(file);
+    }
+
+    TimerOutput::Scope scope(timer, "Compute error norms");
+    Vector<double>     norm_per_cell_p(triangulation.n_active_cells());
+
+    VectorTools::integrate_difference(mapping,
+                                      dof_handler,
+                                      solution,
+                                      ExactSolution<dim>(),
+                                      norm_per_cell_p,
+                                      QGauss<dim>(fe.degree + 2),
+                                      VectorTools::L2_norm);
+    std::cout << "   L2 error vs exact solution:   "
+              << norm_per_cell_p.l2_norm() << std::endl;
+
+    VectorTools::integrate_difference(mapping,
+                                      dof_handler,
+                                      solution,
+                                      ExactSolution<dim>(),
+                                      norm_per_cell_p,
+                                      QGauss<dim>(fe.degree + 2),
+                                      VectorTools::H1_norm);
+    std::cout << "   H1 error vs exact solution:   "
+              << norm_per_cell_p.l2_norm() << std::endl;
+  }
+
+
+
+  template <int dim>
+  void Step64<dim>::run()
+  {
+    create_grid();
+    {
+      std::cout << std::endl
+                << "====== Running with the basic MappingQGeneric class ====== "
+                << std::endl
+                << std::endl;
+      MappingQGeneric<dim> mapping(fe.degree + 1);
+      setup_system(mapping);
+      assemble_system(mapping);
+      solve();
+      output_results(mapping);
+
+      {
+        TimerOutput::Scope scope(timer, "Compute error estimator");
+        Vector<float> estimated_error_per_cell(triangulation.n_active_cells());
```

Comment:
maybe move into ``output_results``? I know you are not using the output except for the timing, but this way you don't duplicate the code here. Maybe you can also print estimated_error_per_cell.l2_norm() or something?

Reply:
Done.